src/expr/macro/src/gen.rs

```diff
@@ -321,13 +321,14 @@ impl FunctionAttr {
         ) #await_ };
         // handle error if the function returns `Result`
         // wrap a `Some` if the function doesn't return `Option`
+        let error_context = format!("in function \"{}\"", self.name);
         output = match user_fn.return_type_kind {
             // XXX: we don't support void type yet. return null::int for now.
             _ if self.ret == "void" => quote! { { #output; Option::<i32>::None } },
             ReturnTypeKind::T => quote! { Some(#output) },
             ReturnTypeKind::Option => output,
-            ReturnTypeKind::Result => quote! { Some(#output?) },
-            ReturnTypeKind::ResultOption => quote! { #output? },
+            ReturnTypeKind::Result => quote! { Some(#output.context(#error_context)?) },
+            ReturnTypeKind::ResultOption => quote! { #output.context(#error_context)? },
```

Comment:
qq: So the most important change is here right? Could you teach me why `anyhow` is necessary for this?

Reply:
I don't think this is the most important part. It is just a side-effect that `anyhow` makes it easy to add context around the error. 🤡

Reply:
Let me rephrase my points: I mean the most important change is the ability to `.context`, right? But does this require `anyhow`? 🤔 (I mean `anyhow` is the implementation detail, instead of the goal) I remember bugen's `thiserror_ext` also have context.

BTW, the most important user-facing change looks indeed like the function name. 🤣 Can you show me some other exciting examples?

Reply:
> I roughly remember bugen's thiserror_ext also have context.

Well, it seems to be `ContextInto`, instead of `ContextInfo` 🤡

Reply:
From developer's view, they can freely use `anyhow` error and no longer need to consider which `ExprError` to use. 🤡

Reply:
Hmmm, I'm thinking about sth like this might suffice

```rust
struct ContextError {
    contexts: Vec<String>,
}

impl ContextError {
    fn context(self, context: impl Display) -> Self {
        let mut contexts = self.contexts;
        contexts.push(context.to_string());
        Self { contexts }
    }
}

// This trait refers to anyhow
trait Context<T> {
    /// Wrap the error value with additional context.
    fn context<C>(self, context: C) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static;

    /// Wrap the error value with additional context that is evaluated lazily
    /// only once an error does occur.
    fn with_context<C, F>(self, f: F) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
        F: FnOnce() -> C;
}

impl<T> Context<T> for Result<T, ContextError> {
    fn context<C>(self, context: C) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(error) => Err(error.context(context)),
        }
    }

    fn with_context<C, F>(self, context: F) -> Result<T, ContextError>
    where
        C: Display + Send + Sync + 'static,
        F: FnOnce() -> C,
    {
        match self {
            Ok(ok) => Ok(ok),
            Err(error) => Err(error.context(context())),
        }
    }
}

fn foo() -> Result<(), ContextError> {
    Err(ContextError {
        contexts: vec!["foo".to_string()],
    })
}

fn bar() -> Result<(), ContextError> {
    foo().context("wtf").context("woo")
}
```

Reply:
Oh, this is just similar to tygg's comment above

> A simple solution is Box<dyn Error> with some syntax sugur (e.g. context).

Reply:
FYI, the implementation of `anyhow::Context::context` uses an anonymous error type to provide the context and put the original error as its `source`, in which way the source chain is correctly maintained.

https://docs.rs/anyhow/latest/src/anyhow/error.rs.html#305-308

Reply:
Actually I view these expr errors as leaf errors. So just feel some string stuff is enough.